### PR TITLE
Force size option added for zvol creation in Middlewared

### PR DESCRIPTION
**Pull request by sonicaj**
Ticket: #33900

### CODE REVIEW

File: src/middlewared/middlewared/plugins/pool.py [hunk 1]
```diff
@@ -619,6 +620,49 @@ def transform(dataset):
 
         verrors = ValidationErrors()
         await self.__common_validation(verrors, 'pool_dataset_create', data, 'CREATE')
+
+        if '/' not in data['name']:
+            verrors.add('pool_dataset_create.name', 'You need a full name, e.g. pool/newdataset')
+        else:
+            parent = await self.middleware.call(
+                'zfs.dataset.query',
+                [('id', '=', data['name'].split('/')[0])]
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: Thats not the parent, you should be using rsplit

File: src/middlewared/middlewared/plugins/pool.py [hunk 2]
```diff
@@ -619,6 +620,49 @@ def transform(dataset):
...
+                )
+            else:
+                parent = parent[0]
+
+                if data['type'] == 'VOLUME' and 'volsize' in data:
+
+                    avail_mem = int(parent['properties']['available']['rawvalue'])
+
+                    if (
+                        data['volsize'] > (avail_mem * 0.80) and
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: You must also do this validation on update, making that not the right math, you must discount the current used size of the zvol.

File: src/middlewared/middlewared/plugins/pool.py [hunk 3]
```diff
@@ -696,9 +718,12 @@ def add(attr):
 
         dataset = await self.middleware.call('pool.dataset.query', [('id', '=', id)])
         if not dataset:
-            verrors.add('id', f'{id} does not exist', errno.ENOENT)
+            verrors.add('pool_dataset_update.id', f'{id} does not exist', errno.ENOENT)
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: You should revert this back I think. `id` is a main schema, not within `pool_dataset_update`

File: src/middlewared/middlewared/plugins/pool.py [hunk 4]
```diff
@@ -755,6 +793,25 @@ def add(attr):
                 if i in data:
                     verrors.add(f'{schema}.{i}', 'This field is not valid for VOLUME')
 
+            if 'volsize' in data and parent:
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: `parent` is always true, right?

File: src/middlewared/middlewared/plugins/pool.py [hunk 5]
```diff
@@ -618,7 +619,28 @@ def transform(dataset):
...
+                data['type'] == 'VOLUME' and
+                all(k in data for k in ('volsize', 'volblocksize'))
+            ):
+
+                if data['volblocksize'].isdigit():
+                    block_size = int(data['volblocksize'])
+                else:
+                    block_size = int(data['volblocksize'][:-1]) * 1024
+
+                if data['volsize'] % block_size:
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: Why are we only checking this on creation?

File: src/middlewared/middlewared/plugins/pool.py [hunk 6]
```diff
@@ -696,9 +718,12 @@ def add(attr):
 
         dataset = await self.middleware.call('pool.dataset.query', [('id', '=', id)])
         if not dataset:
-            verrors.add('id', f'{id} does not exist', errno.ENOENT)
+            verrors.add('pool_dataset_update.id', f'{id} does not exist', errno.ENOENT)
         else:
             data['type'] = dataset[0]['type']
+            data['name'] = dataset[0]['name']
+            if data['type'] == 'VOLUME':
+                data['volsize'] = data['volsize'] if data.get('volsize') else dataset[0]['volsize']['rawvalue']
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: Hmm, that will set the property again if its already set, I think we should avoid defining volsize

File: src/middlewared/middlewared/plugins/pool.py (hunk 4 above)
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/pool.py]: Nope, if mode is create then it's possible that it might not be true

Comment (william-gr) [reply src/middlewared/middlewared/plugins/pool.py]: Care to elaborate? How can it ever not be true?

Comment (william-gr) [reply src/middlewared/middlewared/plugins/pool.py]: Let me add, if you're creating a dataset, the main dataset will always exist, e.g. `tank`, you cannot create `tank/foo` if `tank` does not exist.

**Pull request merged**